### PR TITLE
HttpMethod.Head does not work on some sites

### PR DESCRIPTION
**Pull request by AlexRadch**

### CODE REVIEW

**Comment by Turnerj** [comment]
Not sure what error you are hitting but HTTP Head does work and would work faster than GET as we don't need the contents in the DiscoverSitemaps method.

Here is what Wikipedia says about the HEAD method: "The HEAD method asks for a response identical to that of a GET request, but without the response body. This is useful for retrieving meta-information written in response headers, without having to transport the entire content."

**Comment by Turnerj** [comment]
To further add to that, the HTTP/1.1 specification says general purpose web servers must support GET and HEAD: https://tools.ietf.org/html/rfc2616#section-5.1.1

**Comment by AlexRadch** [comment]
> Not sure what error you are hitting but HTTP Head does work and would work faster than GET as we don't need the contents in the DiscoverSitemaps method.
> 
> Here is what Wikipedia says about the HEAD method: "The HEAD method asks for a response identical to that of a GET request, but without the response body. This is useful for retrieving meta-information written in response headers, without having to transport the entire content."

I know how it should be by standards and etc.

In reality I have sites that do not respond on a HEAD requests. And accordingly, an empty list of sitemap files is returned for them. They respond on a GET request, and then the list of sitemap files is not empty.

Alternatively, you can add an option to select which method to check that the sitemap file exists. But I think that the gain is so insignificant that it is easier for all sites to perform a GET request.

**Comment by Turnerj** [comment]
Ahhh, thanks for explaining it. Out of curiosity, what is returned from the server when HEAD is used? Does it return a 405 Method Not Allowed?

I'm thinking by default it should try HEAD and if it fails for particular reasons, we fall back to GET. The reason for this is that some site maps are huge which can lead to both network and memory overheads while also putting additional load on the server.

**Comment by AlexRadch** [comment]
> Ahhh, thanks for explaining it. Out of curiosity, what is returned from the server when HEAD is used? Does it return a 405 Method Not Allowed?

I checked one. It response
```
{StatusCode: 400, ReasonPhrase: 'Bad Request', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Wed, 29 Jul 2020 04:40:33 GMT
  Connection: keep-alive
  X-Frame-Options: deny
  CF-Cache-Status: DYNAMIC
  cf-request-id: 043a7852300000498d6401c200000001
  Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
  Server: cloudflare
  CF-RAY: 5ba42996bdd7498d-DME
  Content-Type: application/json
}}
```

**Comment by AlexRadch** [comment]
> 
> I'm thinking by default it should try HEAD and if it fails for particular reasons, we fall back to GET. The reason for this is that some site maps are huge which can lead to both network and memory overheads while also putting additional load on the server.

Yes. But you do not read content so only small part of file will be sent.

**Comment by Turnerj** [comment]
Interesting. Could treat any 4xx error (except 404) as reason to fallback from HEAD to GET. That way we get the best of both worlds - perf when we can plus compatibility for servers without HEAD.

**Comment by AlexRadch** [comment]
> 
> 
> Interesting. Could treat any 4xx error (except 404) as reason to fallback from HEAD to GET. That way we get the best of both worlds - perf when we can plus compatibility for servers without HEAD.

I can change pull request to such way.

File: src/TurnerSoftware.SitemapTools/SitemapQuery.cs [hunk 1]
```diff
@@ -63,6 +63,11 @@ public SitemapQuery(HttpClient client)
 			HttpClient = client;
 		}
 
+		///// <summary>
```
Comment (Turnerj) [inline src/TurnerSoftware.SitemapTools/SitemapQuery.cs:66]: Nit: Remove the commented out code

File: src/TurnerSoftware.SitemapTools/SitemapQuery.cs [hunk 2]
```diff
@@ -94,12 +99,23 @@ public async Task<IEnumerable<Uri>> DiscoverSitemapsAsync(string domainName, Can
 
 				try
 				{
+					//var requestMessage = new HttpRequestMessage(IsHeadMethodUnsupported ? HttpMethod.Get : HttpMethod.Head, uri);
```
Comment (Turnerj) [inline src/TurnerSoftware.SitemapTools/SitemapQuery.cs:102]: Nit: Remove the commented out code

File: src/TurnerSoftware.SitemapTools/SitemapQuery.cs [hunk 3]
```diff
@@ -94,12 +99,23 @@ public async Task<IEnumerable<Uri>> DiscoverSitemapsAsync(string domainName, Can
...
 						result.Add(uri);
+						continue;
+					}
+
+					if ((int)response.StatusCode >= 400 && (int)response.StatusCode < 500 && response.StatusCode != HttpStatusCode.NotFound)
+					{
+						requestMessage = new HttpRequestMessage(HttpMethod.Get, uri);
+						response = await HttpClient.SendAsync(requestMessage, cancellationToken);
+
+						if (response.IsSuccessStatusCode)
```
Comment (Turnerj) [inline src/TurnerSoftware.SitemapTools/SitemapQuery.cs:117]: Nit: Use braces for single-line if-statements

**Comment by Turnerj** [comment]
Hey - just wanted to apologise that I didn't do a release with your fix earlier. I do appreciate the time and effort you put into fixing this and, as part of some other changes I was doing, have released this as v0.7.0.

**Pull request merged**